File: packages/react-native-nitro-modules/android/src/main/java/com/margelo/nitro/core/HybridObject.kt
package com.margelo.nitro.core

import androidx.annotation.CallSuper
import androidx.annotation.Keep
import com.facebook.jni.HybridData
import com.facebook.proguard.annotations.DoNotStrip
import dalvik.annotation.optimization.FastNative

/**
 * A base class for all Kotlin-based HybridObjects.
 */
@Keep
@DoNotStrip
abstract class HybridObject {
  /**
   * Get the memory size of the Kotlin instance (plus any external heap allocations),
   * in bytes.
   *
   * Override this to allow tracking heap allocations such as buffers or images,
   * which will help the JS GC be more efficient in deleting large unused objects.
   *
   * @example
   * ```kotlin
   * override val memorySize: ULong
   *   get() {
   *     val imageSize = this.bitmap.bytesPerRow * this.bitmap.height
   *     return imageSize
   *   }
   * ```
   */
  @get:DoNotStrip
  @get:Keep
  open val memorySize: Long
    get() = 0L

  /**
   * Eagerly- (and manually-) dispose all native resources this `HybridObject` holds.
   * This method can only be manually called from JS using `dispose()`.
   *
   * If this method is never manually called, a `HybridObject` is expected to disposes its
   * resources as usual via the object's destructor (`~HybridObject()`, `deinit` or `finalize()`).
   *
   * By default, this method only destroys the connected C++ part.
   * It can be overridden to perform actual disposing/cleanup if required.
   * This method must not throw.
   */
  @DoNotStrip
  @Keep
  @CallSuper
  open fun dispose() {
    synchronized(this) {
      cxxPart?.destroy()
      cxxPart = null
    }
  }

  /**
   * Get a string representation of this `HybridObject` - useful for logging or debugging.
   */
  @DoNotStrip
  @Keep
  override fun toString(): String {
    val ownName = this::class.simpleName
    return "[HybridObject $ownName]"
  }

  /**
   * Override this method for each class in the inheritance
   * chain to connect it to a different C++ class.
   */
  protected open fun createCxxPart(): CxxPart {
    return CxxPart(this)
  }

  /**
   * Java `HybridObject` owns a (lazy) reference to Java `CxxPart`,
   * which itself owns the C++ `CxxPart`.
   * The C++ `CxxPart` must NOT have a reference back to Java,
   * otherwise a cyclic reference outside of JVM is created (= mem leak)
   */
  private var cxxPart: CxxPart? = null

  @Suppress("unused")
  @DoNotStrip
  @Keep
  private fun getCxxPart(): CxxPart {
    synchronized(this) {
      val currentCxxPart = this.cxxPart
      if (currentCxxPart != null) {
        return currentCxxPart
      } else {
        val newCxxPart = createCxxPart()
        this.cxxPart = newCxxPart
        return newCxxPart
      }
    }
  }

  @Keep
  @DoNotStrip
  @Suppress("KotlinJniMissingFunction")
  protected open class CxxPart(
    @Keep
    @DoNotStrip
    val javaPart: HybridObject,
  ) {
    @DoNotStrip
    @Keep
    private var mHybridData: HybridData

    init {
      mHybridData = initHybrid()
    }

    fun destroy() {
      mHybridData.resetNative()
    }

    @FastNative
    protected open external fun initHybrid(): HybridData
  }
}

File: packages/react-native-nitro-modules/ios/core/HybridObject.swift
//
//  HybridObject.swift
//  NitroModules
//
//  Created by Marc Rousavy on 23.07.24.
//

/// A base protocol for all Swift-based Hybrid Objects.
public protocol HybridObject: AnyObject {
  /**
   * Get the memory size of the Swift instance (plus any external heap allocations),
   * in bytes.
   *
   * Override this to allow tracking heap allocations such as buffers or images,
   * which will help the JS GC be more efficient in deleting large unused objects.
   *
   * @example
   * ```swift
   * var memorySize: Int {
   *   let imageSize = self.uiImage.bytesPerRow * self.uiImage.height
   *   return imageSize
   * }
   * ```
   */
  var memorySize: Int { get }

  /**
   * Eagerly- (and manually-) dispose all native resources this `HybridObject` holds.
   * This method can only be manually called from JS using `dispose()`.
   *
   * If this method is never manually called, a `HybridObject` is expected to disposes its
   * resources as usual via the object's destructor (`~HybridObject()`, `deinit` or `finalize()`).
   *
   * By default, this method does nothing. It can be overridden to perform actual disposing/cleanup
   * if required.
   */
  func dispose()

  /**
   * Get a string representation of this `HybridObject` - useful for logging or debugging.
   */
  func toString() -> String
}

extension HybridObject {
  // By default, this returns `0`.
  public var memorySize: Int { return 0 }
  // By default, this does nothing.
  public func dispose() {}
  // By default, this just returns `[HybridObject]`
  public func toString() -> String {
    return "[HybridObject object]"
  }
}
